src/Analyzers/AnalyzerOptionExtensions.cs

```diff
@@ -0,0 +1,118 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Linq;
+
+namespace Microsoft.CodeAnalysis.Diagnostics
+{
+    internal static class AnalyzerOptionsExtensions
+    {
+        private const string DotnetAnalyzerDiagnosticPrefix = "dotnet_analyzer_diagnostic";
+        private const string CategoryPrefix = "category";
+        private const string SeveritySuffix = "severity";
+
+        private const string DotnetAnalyzerDiagnosticSeverityKey = DotnetAnalyzerDiagnosticPrefix + "." + SeveritySuffix;
+
+        private static string GetCategoryBasedDotnetAnalyzerDiagnosticSeverityKey(string category)
+            => $"{DotnetAnalyzerDiagnosticPrefix}.{CategoryPrefix}-{category}.{SeveritySuffix}";
+
+        /// <summary>
+        /// Tries to get configured severity for the given <paramref name="descriptor"/>
+        /// for the given <paramref name="tree"/> from bulk configuration analyzer config options, i.e.
+        ///     'dotnet_analyzer_diagnostic.category-%RuleCategory%.severity = %severity%'
+        ///         or
+        ///     'dotnet_analyzer_diagnostic.severity = %severity%'
+        /// </summary>
+        public static bool TryGetSeverityFromBulkConfiguration(
+            this AnalyzerOptions? analyzerOptions,
+            SyntaxTree tree,
```

Comment:
There was a long discussion today about moving this to live on compilation options instead of the Syntax tree. We can fixup this once its done but it would be nice if roslyn just had a public API that could do this.

Reply:
Yup, I've discussed with Manish about how nice such an API would be.